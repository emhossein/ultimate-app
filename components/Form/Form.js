import React, { useState } from 'react';
import { Button, Paragraph } from '..';
import { useForm } from 'react-hook-form';
import styled from '@emotion/styled';
import { useTheme } from '@emotion/react';

const Form = () => {
  const theme = useTheme();

  const [sent, setSent] = useState(false);

  const submitHandler = data => {
    console.log(data);
    setSent(true);
  };
  const {
    register,
    formState: { errors },
    handleSubmit,
  } = useForm();

  const Wrapper = styled.div`
    width: 100%;
    form {
      display: flex;
      flex-direction: column;
      align-items: center;
      justify-content: center;
      width: 100%;

      input,
      textarea,
      button {
        width: 30%;
        padding: 1.2rem;
        margin-bottom: 0.5rem;
        border-radius: ${theme.borderRadius[3]};
        border: ${theme.border[1]} solid ${theme.borderColor.primary};
        outline: 0;
        font-family: 'Josefin Sans', sans-serif;
      }

      button {
        margin: 0;
      }

      p {
        text-align: center;
        line-height: 1.2;
        margin: 0;
        color: #929292;

        &:last-child {
          margin-top: 1rem;
        }
      }

      .sent {
        margin-top: 1rem;
        color: #000;
      }
    }
  `;

  return (
    <Wrapper className="App">
      <form onSubmit={handleSubmit(submitHandler)}>
        <input
          type="text"
          placeholder="Your Name"
          {...register('firstName', {
            required: 'this is a required',
          })}
        />

        {errors.firstName && <p>please enter your name!</p>}

        <input
          placeholder="Your Email"
          type="text"
          {...register('email', {
            required: 'this is required',
            pattern: {
              value: /^[a-zA-Z0-9.!#$%&â€™*+/=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\.[a-zA-Z0-9-]+)*$/,
              message: 'Invalid email address',
            },
          })}
        />
        {errors.email && <p>please enter your email!</p>}

        <textarea
          name="textarea"
          id="textarea"
          cols="30"
          rows="10"
          placeholder="write something"
          {...register('textarea', {
            required: true,
          })}
        />
        {errors.textarea && <Paragraph>Please write something!</Paragraph>}
        <Button type="submit">Submit</Button>
        {sent && <Paragraph>Message has been sent!</Paragraph>}
      </form>
    </Wrapper>
  );
};

export default Form;
