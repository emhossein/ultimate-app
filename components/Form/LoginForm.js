import React, { useState } from 'react';
import { Button, Paragraph } from '..';
import { useForm } from 'react-hook-form';
import styled from '@emotion/styled';
import { useTheme } from '@emotion/react';
import { useDispatch } from 'react-redux';
import { useRouter } from 'next/router';
import { getAuthStatus } from '../../slices/authSlice';

const LoginForm = ({ user, loading }) => {
  const router = useRouter();
  const theme = useTheme();
  const dispatch = useDispatch();

  const [error, setError] = useState(false);

  const {
    register,
    handleSubmit,
    formState: { errors },
    watch,
  } = useForm();

  const email = watch('email', false);
  const password = watch('password', false);

  const exactEmail = user.map(use => use.email);
  const exactPassword = user.map(use => use.password);

  const submitHandler = async data => {
    if (!exactEmail.includes(email) || !exactPassword.includes(password)) {
      console.log('Incorrect username or password.');
      setError(true);
      return;
    } else {
      dispatch(getAuthStatus());
      router.push('/');
      setError(false);
    }
  };

  const Wrapper = styled.div`
    width: 100%;
    form {
      display: flex;
      flex-direction: column;
      align-items: center;
      justify-content: center;
      width: 100%;

      input,
      textarea,
      button {
        width: 30%;
        padding: 1.2rem;
        margin-bottom: 0.5rem;
        border-radius: ${theme.borderRadius[3]};
        border: ${theme.border[1]} solid ${theme.borderColor.primary};
        outline: 0;
        font-family: 'Josefin Sans', sans-serif;
      }

      button {
        margin: 0;
      }

      p {
        margin: 0 auto;
        line-height: 1.2;
        margin: 0;
        color: #929292;
      }

      p:last-child {
        margin: 1rem 0;
        color: red;
      }
    }
  `;

  return (
    <Wrapper>
      <form onSubmit={handleSubmit(submitHandler)}>
        <input
          placeholder="Your Email"
          type="text"
          {...register('email', {
            required: 'this is required',
            pattern: {
              value: /^[a-zA-Z0-9.!#$%&â€™*+/=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\.[a-zA-Z0-9-]+)*$/,
              message: 'Invalid email address',
            },
          })}
        />
        {errors.email && <Paragraph>please enter your email!</Paragraph>}

        <input
          type="password"
          placeholder="Password"
          {...register('password', {
            required: 'this is required',
            minLength: {
              value: 6,
              message: 'at least 6 character',
            },
          })}
        />

        {errors.password && <Paragraph>please enter your password!</Paragraph>}

        <Button>{loading ? <div className="loading" /> : 'LOGIN'}</Button>
        {error && <Paragraph>Incorrect username or password.</Paragraph>}
      </form>
    </Wrapper>
  );
};

export default LoginForm;
