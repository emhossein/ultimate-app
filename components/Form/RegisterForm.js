import React, { useState } from 'react';
import { Button } from '..';
import { useForm } from 'react-hook-form';
import styled from '@emotion/styled';
import { useTheme } from '@emotion/react';
import { useDispatch, useSelector } from 'react-redux';
import { useRouter } from 'next/router';
import { getAuthStatus } from '../../slices/authSlice';

const Form = () => {
  const router = useRouter();
  const theme = useTheme();
  const dispatch = useDispatch();
  const { loading } = useSelector(state => state.auth);

  const {
    register,
    handleSubmit,
    formState: { errors },
  } = useForm();

  const submitHandler = async data => {
    dispatch(getAuthStatus());
    router.push('/');
  };

  const Wrapper = styled.div`
    width: 100%;
    form {
      display: flex;
      flex-direction: column;
      align-items: center;
      justify-content: center;
      width: 100%;

      input,
      textarea,
      button {
        width: 30%;
        padding: 1.2rem;
        margin-bottom: 0.5rem;
        border-radius: ${theme.borderRadius[3]};
        border: ${theme.border[1]} solid ${theme.borderColor.primary};
        outline: 0;
        font-family: 'Josefin Sans', sans-serif;
      }

      button {
        margin: 0;
      }

      p {
        text-align: center;
        line-height: 1.2;
        margin: 0;
        color: #929292;
      }
    }
  `;

  return (
    <Wrapper>
      <form onSubmit={handleSubmit(submitHandler)}>
        <input
          type="text"
          placeholder="Your Name"
          {...register('firstName', {
            required: 'this is a required',
          })}
        />

        {errors.firstName && <p>please enter your name!</p>}
        <input
          placeholder="Your Email"
          type="text"
          {...register('email', {
            required: 'this is required',
            pattern: {
              value: /^[a-zA-Z0-9.!#$%&â€™*+/=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\.[a-zA-Z0-9-]+)*$/,
              message: 'Invalid email address',
            },
          })}
        />
        {errors.email && <p>please enter your email!</p>}

        <input
          type="password"
          placeholder="Password"
          {...register('password', {
            required: 'this is required',
            minLength: {
              value: 6,
              message: 'at least 6 character',
            },
          })}
        />

        {errors.password && <p>please enter your password!</p>}

        <Button>{loading ? <div className="loading" /> : 'REGISTER'}</Button>
      </form>
    </Wrapper>
  );
};

export default Form;
