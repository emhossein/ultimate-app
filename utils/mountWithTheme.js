import React from 'react';
import renderer from 'react-test-renderer';
import { ThemeProvider } from '@emotion/react';
import theme from '../config/theme';

const mountWithTheme = ({ children }) => {
  return renderer.create(
    <ThemeProvider theme={theme}>{children}</ThemeProvider>
  );
};

export default mountWithTheme;
