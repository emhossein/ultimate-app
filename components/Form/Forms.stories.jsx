import React from 'react';

import Form from './Form';

export default {
  title: 'Forms',
};

export const contactForm = () => <Form />;
