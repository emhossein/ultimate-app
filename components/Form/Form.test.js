/**
 * @jest-environment jsdom
 */

import React from 'react';
import { createSerializer } from '@emotion/jest';
import mountWithTheme from '../../utils/mountWithTheme';
import Form from './Form';
import { LoginForm, RegisterForm } from '..';

expect.addSnapshotSerializer(createSerializer());

describe('Forms', () => {
  test('Form Renders Correctly', () => {
    const component = mountWithTheme(<Form />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  test('Login Form Renders Correctly', () => {
    const component = mountWithTheme(<LoginForm />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  test('Register Form Renders Correctly', () => {
    const component = mountWithTheme(<RegisterForm />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
