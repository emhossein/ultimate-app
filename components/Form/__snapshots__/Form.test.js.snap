// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Forms Form Renders Correctly 1`] = `null`;

exports[`Forms Login Form Renders Correctly 1`] = `null`;

exports[`Forms Register Form Renders Correctly 1`] = `null`;

exports[`Typography H1 Renders Correctly 1`] = `null`;

exports[`Typography H2 Renders Correctly 1`] = `null`;

exports[`Typography H3 Renders Correctly 1`] = `null`;
